packages/inertia/src/inertia.js

```diff
@@ -117,25 +132,31 @@ export default {
     Progress.increment()
     return Promise.resolve(this.resolveComponent(page.component)).then(component => {
       if (visitId === this.visitId) {
-        preserveState = typeof preserveState === 'function' ? preserveState(page.props) : preserveState
-        preserveScroll = typeof preserveScroll === 'function' ? preserveScroll(page.props) : preserveScroll
-
         this.version = page.version
         this.setState(page, replace, preserveState)
-        this.updatePage(component, page.props, { preserveState })
-        this.setScroll(preserveScroll)
+        this.updatePage(component, page.props, { preserveState }).then(() => {
+          let scrollRegions = document.querySelectorAll('[scroll-region]')
+
+          scrollRegions.forEach(region => {
+            region.addEventListener('scroll', this.saveScrollPositions)
```

Comment:
You could also do the `debounce` here instead of in the function declaration.
```suggestion
            region.addEventListener('scroll', debounce(this.saveScrollPositions, 100))
```

Reply:
The issue here is that the old event listeners are only remove if you use the exact same function reference. We need the old event listeners to be removed when you navigate to a new page, otherwise you'll eventually get hundreds of them.